### PR TITLE
[NFC] Add SILGenCleanup::completeOSSLifetimes

### PR DESCRIPTION
**Pull request by atrick**
Presently under -enable-ossa-complete-lifetimes.

This allows SILGen to skip OSSA cleanups, for example at dead-end
blocks.

Long term, we may remove OSSA cleanups from SILGen entirely (except
for lexical borrow scopes). This changes lets us experiment with that
option.

### CODE REVIEW

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
@swift-ci test

**Comment by atrick** [comment]
@swift-ci smoke test

**Comment by atrick** [comment]
PR testing is blocked by
```
Reflection/Sources/_Runtime/Utils/PtrAuth.swift:201:7: error: cannot convert value of type 'UInt64' to expected argument type 'Builtin.Int64'
      UInt64(Builtin.typePtrAuthDiscriminator(
      ^
```

**Review by nate-chandler — COMMENTED** [review]
LGTM!

File: lib/SIL/Utils/OSSALifetimeCompletion.cpp [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+//
+//===----------------------------------------------------------------------===//
+///
+/// OSSA lifetime completion adds lifetime ending instructions to make
+/// linear lifetimes complete.
+///
+/// Interior liveness handles the following cases naturally:
+///
+/// When completing the lifetime if the initial value, %v1, transitively
+/// include all dominated reborrows. %phi1 in this example:
```
[the remark this answers is not in the document]
Comment (atrick) [reply lib/SIL/Utils/OSSALifetimeCompletion.cpp:19]: Fixed in https://github.com/apple/swift/pull/64051

File: lib/SIL/Utils/OSSALifetimeCompletion.cpp [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+
+#include "swift/SIL/OSSALifetimeCompletion.h"
+#include "swift/SIL/SILBuilder.h"
+#include "swift/SIL/SILInstruction.h"
+
+using namespace swift;
+
+static SILInstruction *endOSSALifetime(SILValue value, SILBuilder &builder) {
+  auto loc =
+    RegularLocation::getAutoGeneratedLocation(builder.getInsertionPointLoc());
```
[the remark this answers is not in the document]
Comment (atrick) [reply lib/SIL/Utils/OSSALifetimeCompletion.cpp:60]: We use `RegularLocation::getAutoGeneratedLocation` everywhere for compiler-generated destroys. I really don't know if `CleanupLocation` will work once we move the destroy somewhere else. CleanupLocation seems to be for proper cleanups generated by the frontend. (@adrian-prantl ). At any rate, if we're doing it wrong, we should probably change our approach everywhere. Until then, I'll be consistent.

[the remark this answers is not in the document]
Comment (nate-chandler) [reply lib/SIL/Utils/OSSALifetimeCompletion.cpp:60]: It looks like everywhere that's done we may run afoul of the debug scope hole checking.  The code that checks for them https://github.com/apple/swift/blob/main/lib/SIL/Verifier/SILVerifier.cpp#L6216-L6217 does skip over instructions whose locations are cleanups but not those that are autogenerated.  I've encountered this problem before: https://github.com/apple/swift/pull/62533 .

**Pull request merged**